core-backend/common/src/memory.rs

```diff
@@ -47,22 +50,52 @@ pub enum ProcessAccessError {
 
 #[derive(Debug, Clone, derive_more::From)]
 pub enum MemoryAccessError {
-    #[from]
     Memory(MemoryError),
-    #[from]
+    ProcessAccess(ProcessAccessError),
     RuntimeBuffer(RuntimeBufferSizeError),
     // TODO: remove #2164
     Decode,
-    GasLimitExceeded,
-    GasAllowanceExceeded,
 }
 
-impl From<ProcessAccessError> for MemoryAccessError {
-    fn from(err: ProcessAccessError) -> Self {
-        match err {
-            ProcessAccessError::OutOfBounds => MemoryError::AccessOutOfBounds.into(),
-            ProcessAccessError::GasLimitExceeded => Self::GasLimitExceeded,
-            ProcessAccessError::GasAllowanceExceeded => Self::GasAllowanceExceeded,
+impl BackendExternalitiesError for MemoryAccessError {
+    fn into_termination_reason(self) -> TerminationReason {
+        match self {
+            MemoryAccessError::ProcessAccess(ProcessAccessError::OutOfBounds)
+            | MemoryAccessError::Memory(MemoryError::AccessOutOfBounds) => {
+                TrapExplanation::UnrecoverableExt(
+                    UnrecoverableMemoryError::AccessOutOfBounds.into(),
+                )
+                .into()
+            }
+            MemoryAccessError::RuntimeBuffer(RuntimeBufferSizeError) => {
+                TrapExplanation::UnrecoverableExt(
+                    UnrecoverableMemoryError::RuntimeAllocOutOfBounds.into(),
+                )
+                .into()
+            }
+            MemoryAccessError::ProcessAccess(ProcessAccessError::GasLimitExceeded) => {
+                TrapExplanation::GasLimitExceeded.into()
+            }
+            MemoryAccessError::ProcessAccess(ProcessAccessError::GasAllowanceExceeded) => {
+                ActorTerminationReason::GasAllowanceExceeded
+            }
+            MemoryAccessError::Decode => unreachable!(),
+        }
+        .into()
+    }
+
+    fn into_run_fallible_error(self) -> RunFallibleError {
+        match self {
+            MemoryAccessError::Memory(MemoryError::AccessOutOfBounds) => {
+                RunFallibleError::FallibleExt(FallibleMemoryError::AccessOutOfBounds.into())
+            }
+            MemoryAccessError::ProcessAccess(ProcessAccessError::OutOfBounds) => {
+                RunFallibleError::FallibleExt(FallibleMemoryError::AccessOutOfBounds.into())
+            }
+            MemoryAccessError::RuntimeBuffer(RuntimeBufferSizeError) => {
+                RunFallibleError::FallibleExt(FallibleMemoryError::RuntimeAllocOutOfBounds.into())
+            }
```

Comment:
are they really fallible? I mean these error aren't handled during processing fallible func result. Better map these variants to `unreachable` (if it's unreachable).

Reply:
@grishasobol 
How shall we treat `RuntimeAllocOutOfBounds` & `AccessOutOfBounds` erros? Are they "fallible"?

Reply:
> are they really fallible?

Yes. Fallible sys-calls use `read_as` and `read_decoded` methods so they return `RuntimeAllocOutOfBounds & AccessOutOfBounds`.

Reply:
All right here for me